pfam_experiment.py

```diff
@@ -46,186 +48,439 @@
+    """Creates representation model (encoder --> lens --> predictor) architecture."""
+
+    num_families = len(family_ids)
+
+    if use_transformer:
+
+        if use_bert:
+            model_cls = models.FlaxBERT
+        else:
+            model_cls = models.FlaxLM
+
+        if encoder_fn_params is not None:
+            pretrained_transformer_params = encoder_fn_params
+        else:
+            if FLAGS.restore_transformer_dir is not None:
+                pretrained_transformer_params = load_transformer_params(
+                    FLAGS.restore_transformer_dir, model_cls)
+            else:
+                pretrained_transformer_params = None
+
+        model = create_transformer_representation_model(
+            transformer_kwargs=encoder_fn_kwargs,
+            reduce_fn=reduce_fn,
+            reduce_fn_kwargs=reduce_fn_kwargs,
+            num_categories=pfam_num_categories,
+            output_features=num_families,
+            bidirectional=FLAGS.use_bert,
+            output=output,
+            key=random.PRNGKey(FLAGS.random_key),
+            encoder_fn_params=pretrained_transformer_params,
+            reduce_fn_params=reduce_fn_params,
+            predict_fn_params=predict_fn_params)
+
+    else:
+        model = create_representation_model(
+            encoder_fn=encoder_fn,
+            encoder_fn_kwargs=encoder_fn_kwargs,
+            reduce_fn=reduce_fn,
+            reduce_fn_kwargs=reduce_fn_kwargs,
+            num_categories=pfam_num_categories,
+            output_features=num_families,
+            output=output,
+            key=random.PRNGKey(FLAGS.random_key),
+            encoder_fn_params=encoder_fn_params,
+            reduce_fn_params=reduce_fn_params,
+            predict_fn_params=predict_fn_params)
+
+    return model
+
+
+def measure_nearest_neighbor_performance(accuracy_label, encoder,
+                                         family_accessions, batch_size,
+                                         train_samples, shuffle_seed,
+                                         sample_random_state):
+    """Measures nearest neighbor classification performance and updates datum."""
+
+    results = pfam_nearest_neighbors_classification(
+        encoder=encoder,
+        family_accessions=family_accessions,
+        batch_size=batch_size,
+        train_samples=train_samples,
+        shuffle_seed=shuffle_seed,
+        sample_random_state=sample_random_state,
+        data_partitions_dirpath=FLAGS.data_partitions_dirpath,
+        gcs_bucket=FLAGS.gcs_bucket)[0]
+
+    accuracy = results['1-nn accuracy']
+
+    accuracy_dict = {accuracy_label: accuracy}
+
+    return accuracy_dict
+
+
+def parameter_update(model, params):
+    """Updates a model's parameters using a parameters dictionary."""
+
+    assert (
+        model.params.keys() == params.keys()), 'Model parameters do not match!'
+    
+    for layer in model.params.keys():
+        model.params[layer] = params[layer]
```

Comment:
rather than assuming that deepcopy was called upstream, it would be a better practice to do the copy inside here, since this is the line where you're assigning to a reference.